x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/ClusterDeprecationChecks.java

```diff
@@ -167,4 +169,24 @@ static DeprecationIssue checkPollIntervalTooLow(ClusterState state) {
         }
         return null;
     }
+
+    static DeprecationIssue checkTemplatesWithMultipleTypes(ClusterState state) {
+        Set<String> templatesWithMultipleTypes = new HashSet<>();
+        state.getMetadata().getTemplates().forEach((templateCursor) -> {
+            String templateName = templateCursor.key;
+            ImmutableOpenMap<String, CompressedXContent> mappings = templateCursor.value.mappings();
+            if (mappings != null && mappings.size() > 1) {
+                templatesWithMultipleTypes.add(templateName);
+            }
+        });
+        if (templatesWithMultipleTypes.isEmpty()) {
+            return null;
+        }
+        return new DeprecationIssue(DeprecationIssue.Level.CRITICAL,
```

Comment:
Just for my knowledge, does `CRITICAL` mean we'll prevent the upgrade from proceeding?

Reply:
Good question, I've asked in the core-features channel and will try to find out!